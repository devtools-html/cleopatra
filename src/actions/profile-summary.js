/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';
import type { Summary } from '../profile-logic/summarize-profile';
import type { ThreadIndex } from '../types/profile';

export function profileSummaryProcessed(summary: Summary): Action {
  return {
    type: 'PROFILE_SUMMARY_PROCESSED',
    summary,
  };
}

export function expandProfileSummaryThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'PROFILE_SUMMARY_EXPAND',
    threadIndex,
  };
}

export function collapseProfileSummaryThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'PROFILE_SUMMARY_COLLAPSE',
    threadIndex,
  };
}
