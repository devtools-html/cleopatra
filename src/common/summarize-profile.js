import { timeCode } from './time-code';

/**
 * A list of strategies for matching sample names to patterns.
 */
const match = {
  exact: (symbol, pattern) => symbol === pattern,
  prefix: (symbol, pattern) => symbol.startsWith(pattern),
  substring: (symbol, pattern) => symbol.includes(pattern),
  stem: (symbol, pattern) => {
    return symbol === pattern || (symbol && symbol.startsWith(pattern + '('));
  },
};

/**
 * Categories is a list that includes the necessary information to match a sample to
 * a category. This list will need to be adjusted as the engine implementation switches.
 * Each category definition is a tuple that takes the following form:
 *
 * [
 *   matches, // A function that returns true/false for how the pattern should be matched.
 *   pattern, // The pattern that should match the sample name.
 *   category, // The category to finally label the sample.
 * ]
 */
const categories = [
  [match.exact, 'js::RunScript', 'script'],
  [match.stem, 'js::Nursery::collect', 'GC'],
  [match.stem, 'js::GCRuntime::collect', 'GC'],
  [match.prefix, 'mozilla::RestyleManager::', 'restyle'],
  [match.substring, 'RestyleManager', 'restyle'],
  [match.stem, 'PresShell::ProcessReflowCommands', 'layout'],
  [match.prefix, 'nsCSSFrameConstructor::', 'frameconstruction'],
  [match.stem, 'PresShell::DoReflow', 'layout'],
  [match.substring, '::compileScript(', 'script'],

  [match.prefix, 'nsCycleCollector', 'CC'],
  [match.prefix, 'nsPurpleBuffer', 'CC'],
  [match.substring, 'pthread_mutex_lock', 'wait'], // eg __GI___pthread_mutex_lock
  [match.prefix, 'nsRefreshDriver::IsWaitingForPaint', 'paint'], // arguable, I suppose
  [match.stem, 'PresShell::Paint', 'paint'],
  [match.prefix, '__poll', 'wait'],
  [match.prefix, '__pthread_cond_wait', 'wait'],
  [match.stem, 'PresShell::DoUpdateApproximateFrameVisibility', 'layout'], // could just as well be paint
  [match.substring, 'mozilla::net::', 'network'],
  [match.stem, 'nsInputStreamReadyEvent::Run', 'network'],

  // [match.stem, 'NS_ProcessNextEvent', 'eventloop'],
  [match.exact, 'nsJSUtil::EvaluateString', 'script'],
  [match.prefix, 'js::frontend::Parser', 'script.parse'],
  [match.prefix, 'js::jit::IonCompile', 'script.compile.ion'],
  [match.prefix, 'js::jit::BaselineCompiler::compile', 'script.compile.baseline'],

  [match.prefix, 'CompositorBridgeParent::Composite', 'paint'],
  [match.prefix, 'mozilla::layers::PLayerTransactionParent::Read(', 'messageread'],

  [match.prefix, 'mozilla::dom::', 'dom'],
  [match.prefix, 'nsDOMCSSDeclaration::', 'restyle'],
  [match.prefix, 'nsHTMLDNS', 'network'],
  [match.substring, 'IC::update(', 'script.icupdate'],
  [match.prefix, 'js::jit::CodeGenerator::link(', 'script.link'],

  [match.exact, 'base::WaitableEvent::Wait()', 'idle'],
  // TODO - if mach_msg_trap is called by RunCurrentEventLoopInMode, then it
  // should be considered idle time. Add a fourth entry to this tuple
  // for child checks?
  [match.exact, 'mach_msg_trap', 'wait'],

  // Can't do this until we come up with a way of labeling ion/baseline.
  [match.prefix, 'Interpret(', 'script.execute.interpreter'],
];

export function summarizeProfile(profile) {
  return timeCode('summarizeProfile', () => {
    const categories = categorizeThreadSamples(profile);
    const rollingSummaries = calculateRollingSummaries(profile, categories);
    const summaries = summarizeCategories(profile, categories);

    return profile.threads.map((thread, i) => ({
      threadIndex: i,
      threadName: thread.name,
      rollingSummary: rollingSummaries[i],
      summary: summaries[i],
    }));
  });
}

/**
 * Return a function that categorizes a function name. The categories
 * are cached between calls.
 * @returns {function} Function categorizer.
 */
function functionNameCategorizer() {
  const cache = new Map();
  return function functionNameToCategory(name) {
    const existingCategory = cache.get(name);
    if (existingCategory !== undefined) {
      return existingCategory;
    }

    for (const [matches, pattern, category] of categories) {
      if (matches(name, pattern)) {
        cache.set(name, category);
        return category;
      }
    }

    cache.set(name, false);
    return false;
  };
}

/**
 * Given a profile, return a function that categorizes a sample.
 * @param {object} thread Thread from a profile.
 * @return {function} Sample stack categorizer.
 */
function sampleCategorizer(thread) {
  const categorizeFuncName = functionNameCategorizer();

  function computeCategory(stackIndex) {
    if (stackIndex === null) {
      return 'uncategorized';
    }

    const frameIndex = thread.stackTable.frame[stackIndex];
    const implIndex = thread.frameTable.implementation[frameIndex];
    if (implIndex !== null) {
      // script.execute.baseline or script.execute.ion
      return 'script.execute.' + thread.stringTable._array[implIndex];
    }

    const funcIndex = thread.frameTable.func[frameIndex];
    const name = thread.stringTable._array[thread.funcTable.name[funcIndex]];
    const category = categorizeFuncName(name);
    if (category !== false && category !== 'wait') {
      return category;
    }

    const prefixCategory = categorizeSampleStack(thread.stackTable.prefix[stackIndex]);
    if (category === 'wait') {
      if (prefixCategory === 'uncategorized') {
        return 'wait';
      }
      if (prefixCategory.endsWith('.wait') || prefixCategory === 'wait') {
        return prefixCategory;
      }
      return prefixCategory + '.wait';
    }

    return prefixCategory;
  }

  const stackCategoryCache = new Map();
  function categorizeSampleStack(stackIndex) {
    let category = stackCategoryCache.get(stackIndex);
    if (category !== undefined) {
      return category;
    }

    category = computeCategory(stackIndex);
    stackCategoryCache.set(stackIndex, category);
    return category;
  }

  return categorizeSampleStack;
}

/**
 * Count the number of samples in a given category. This will also count subcategories
 * in the case of categories labeled like "script.link", so "script" and "script.link"
 * will each be counted as having a sample.
 * @param {object} summary - Accumulates the counts.
 * @param {string} fullCategoryName - The name of the category.
 * @returns {object} summary
 */
function summarizeSampleCategories(summary, fullCategoryName) {
  const categories = fullCategoryName.split('.');

  while (categories.length > 0) {
    const category = categories.join('.');
    summary[category] = (summary[category] || 0) + 1;
    categories.pop();
  }

  return summary;
}

/**
 * Finalize the summary calculation by attaching percentages and sorting the result.
 * @param {object} summary - The object that summarizes the times of the samples.
 * @return {array} The summary with percentages.
 */
function calculateSummaryPercentages(summary) {
  const rows = Object.entries(summary);

  const sampleCount = rows.reduce((sum, [name, count]) => {
    // Only count the sample if it's not a sub-category. For instance "script.link"
    // is a sub-category of "script".
    return sum + (name.includes('.') ? 0 : count);
  }, 0);

  return rows
    .map(([category, samples]) => {
      const percentage = samples / sampleCount;
      return { category, samples, percentage };
    })
    .sort((a, b) => b.samples - a.samples);
}

function logStacks(samples, maxLogLength = 10) {
  const entries = Object.entries(samples);
  /* eslint-disable no-console */
  console.log(`Top ${maxLogLength} stacks in selected category`);
  const log = typeof console.table === 'function' ? console.table : console.log;
  log(
    entries
      .sort(([, {total: a}], [, {total: b}]) => b - a)
      .slice(0, Math.min(maxLogLength, entries.length))
  );
  /* eslint-enable no-console */
}

function stackToString(stackIndex, thread) {
  const { stackTable, frameTable, funcTable, stringTable } = thread;
  const stack = [];
  let nextStackIndex = stackIndex;
  while (nextStackIndex !== null) {
    const frameIndex = stackTable.frame[nextStackIndex];
    const funcIndex = frameTable.func[frameIndex];
    const name = stringTable._array[funcTable.name[funcIndex]];
    stack.push(name);
    nextStackIndex = stackTable.prefix[nextStackIndex];
  }
  return stack.join('\n');
}

function incrementPerThreadCount(container, key, threadName) {
  const count = container[key] || { total: 0, [threadName]: 0 };
  count.total++;
  count[threadName]++;
  container[key] = count;
}

function countStacksInCategory(profile, summaries, category = 'uncategorized') {
  const uncategorized = {};
  profile.threads.forEach((thread, i) => {
    const threadSummary = summaries[i];
    const { samples } = thread;
    for (let sampleIndex = 0; sampleIndex < samples.length; sampleIndex++) {
      if (threadSummary[sampleIndex] === category) {
        const stringCallStack = stackToString(samples.stack[sampleIndex], thread);
        incrementPerThreadCount(uncategorized, stringCallStack, thread.name);
      }
    }
  });
  return uncategorized;
}

/**
 * Take a profile and return a summary that categorizes each sample, then calculate
 * a summary of the percentage of time each sample was present.
 * @param {array} profile - The current profile.
 * @returns {array} Stacks mapped to categories.
 */
export function categorizeThreadSamples(profile) {
  return timeCode('categorizeThreadSamples', () => {
    const summaries = profile.threads.map(thread => {
      const categorizer = sampleCategorizer(thread);
      return thread.samples.stack.map(categorizer);
    });

    if (process.env.NODE_ENV === 'development') {
      // Change the constant to display the top stacks of a different category.
      const categoryToDump = 'uncategorized';
      const stacks = countStacksInCategory(profile, summaries, categoryToDump);
      console.log(`${Object.keys(stacks).length} stacks labeled '${categoryToDump}'`);
      logStacks(stacks);
    }

    return summaries;
  });
}

/**
 * Take a profile and return a summary that categorizes each sample, then calculate
 * a summary of the percentage of time each sample was present.
 * @param {object} profile - The profile to summarize.
 * @param {object} threadCategories - Each thread's categories for the samples.
 * @returns {object} The summaries of each thread.
 */
export function summarizeCategories(profile, threadCategories) {
  return threadCategories.map(categories => (
      categories.reduce(summarizeSampleCategories, {})
    ))
    .map(calculateSummaryPercentages);
    // Sort the threads based on how many categories they have.
    // .sort((a, b) => Object.keys(b.summary).length - Object.keys(a.summary).length);
}

export function calculateRollingSummaries(profile, threadCategories, segmentCount = 40, rolling = 4) {
  const [minTime, maxTime] = profile.threads.map(thread => {
    return [thread.samples.time[0], thread.samples.time[thread.samples.time.length - 1]];
  })
  .reduce((a, b) => ([
    Math.min(a[0], b[0]),
    Math.max(a[1], b[1]),
  ]));
  const totalTime = maxTime - minTime;
  const segmentLength = totalTime / segmentCount;
  const segmentHalfLength = segmentLength / 2;
  const rollingLength = segmentLength * rolling;
  const rollingHalfLength = segmentLength * rolling / 2;

  return profile.threads.map((thread, threadIndex) => {
    const categories = threadCategories[threadIndex];

    return times(segmentCount, segmentIndex => {
      let samplesInRange = 0;
      const samples = {};

      const rollingMinTime = minTime + (segmentIndex * segmentLength) + segmentHalfLength - rollingHalfLength;
      const rollingMaxTime = rollingMinTime + rollingLength;

      for (let sampleIndex = 0; sampleIndex < thread.samples.time.length; sampleIndex++) {
        const time = thread.samples.time[sampleIndex];
        if (time > rollingMinTime) {
          if (time > rollingMaxTime) {
            break;
          }
          const category = categories[sampleIndex];
          samples[category] = (samples[category] || 0) + 1;
          samplesInRange++;
        }
      }

      return {
        samples,
        percentage: mapObj(samples, count => count / samplesInRange),
      };
    });
  });
}

function times(n, fn) {
  const results = Array(n);
  for (let i = 0; i < n; i++) {
    results[i] = fn(i);
  }
  return results;
}

function mapObj(object, fn) {
  let i = 0;
  const mappedObj = {};
  for (const key in object) {
    if (object.hasOwnProperty(key)) {
      i++;
      mappedObj[key] = fn(object[key], key, i);
    }
  }
  return mappedObj;
}
