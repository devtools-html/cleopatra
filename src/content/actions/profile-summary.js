// @flow
import type { Action } from './types';
import type { Summary } from '../../common/summarize-profile';
import type { ThreadIndex } from '../../common/types/profile';

export function profileSummaryProcessed(summary: Summary): Action {
  return {
    type: 'PROFILE_SUMMARY_PROCESSED',
    summary,
  };
}

export function expandProfileSummaryThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'PROFILE_SUMMARY_EXPAND',
    threadIndex,
  };
}

export function collapseProfileSummaryThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'PROFILE_SUMMARY_COLLAPSE',
    threadIndex,
  };
}
