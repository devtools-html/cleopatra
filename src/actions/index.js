/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
