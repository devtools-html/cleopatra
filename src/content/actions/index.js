// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
