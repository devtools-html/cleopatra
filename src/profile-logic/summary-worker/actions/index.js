/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { summarizeProfile } from '../../../profile-logic/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
