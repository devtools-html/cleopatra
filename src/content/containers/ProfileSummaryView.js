import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { getProfile } from '../reducers/profile-view';
import { getProfileSummaries, getProfileExpandedSummaries } from '../reducers/summary-view';
import SummarizeLineGraph from '../components/SummarizeLineGraph';
import SummarizeProfileHeader from '../components/SummarizeProfileHeader';
import SummarizeProfileExpand from '../components/SummarizeProfileExpand';
import SummarizeProfileThread from '../components/SummarizeProfileThread';
import actions from '../actions';

require('./ProfileSummaryView.css');

const EXPAND_LENGTH = 5;

class ProfileSummaryView extends Component {
  render() {
    const {
      summaries, expanded, threads,
      collapseProfileSummaryThread: collapse,
      expandProfileSummaryThread: expand,
    } = this.props;

    if (summaries) {
      return (
        <div className='summarize-profile'>
          <div className='summarize-profile-inner'>
            {summaries.map(({threadIndex, summary, rollingSummary}) => {
              const { processType, name: threadName } = threads[threadIndex];
              const isExpanded = expanded.has(threadIndex);

              return (
                <div key={threadIndex}>
                  <div className='summarize-profile-table'>
                    <SummarizeProfileHeader threadName={threadName}
                                            processType={processType} />
                    {summary.map((summaryTable, index) => (
                      <SummarizeProfileThread
                        summaryTable={summaryTable}
                        rollingSummary={rollingSummary}
                        isExpanded={isExpanded}
                        index={index}
                        key={summaryTable.category}
                        expandLength={EXPAND_LENGTH} />
                    ))}
                    <SummarizeProfileExpand
                      summary={summary}
                      threadIndex={threadIndex}
                      isExpanded={isExpanded}
                      expand={expand}
                      collapse={collapse}
                      expandLength={EXPAND_LENGTH} />
                    </div>
                </div>
              );
            })}
          </div>
        </div>
      );
    }

    return (
      <div className='summarize-profile'>
        <div className='summarize-profile-inner'>
          {threads.map((thread, threadIndex) => (
            <div key={threadIndex}>
              <div className='summarize-profile-table'>
                <SummarizeProfileHeader threadName={thread.name}
                                        processType={thread.processType} />
                {fill(3, i => (
                  <div className='summarize-profile-row' key={i}>
                    <SummarizeLineGraph />
                    <div className='summarize-profile-details'>
                      <div className='summarize-profile-text'><div className='filler summarize-profile-filler'></div></div>
                      <div className='summarize-profile-numeric'><div className='filler summarize-profile-filler'></div></div>
                      <div className='summarize-profile-numeric'><div className='filler summarize-profile-filler'></div></div>
                    </div>
                  </div>
                ))}
              </div>
            </div>
          ))}
        </div>
      </div>
    );
  }
}

ProfileSummaryView.propTypes = {
  summaries: PropTypes.array,
  expanded: PropTypes.object,
  threads: PropTypes.array,
  collapseProfileSummaryThread: PropTypes.func,
  expandProfileSummaryThread: PropTypes.func,
};

function fill(size, fn) {
  const array = Array(size);
  for (let i = 0; i < size; i++) {
    array[i] = fn(i);
  }
  return array;
}


export default connect(state => {
  return {
    expanded: getProfileExpandedSummaries(state),
    summaries: getProfileSummaries(state),
    threads: getProfile(state).threads,
  };
}, actions)(ProfileSummaryView);
