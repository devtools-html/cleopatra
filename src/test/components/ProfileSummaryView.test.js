/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import React from 'react';
import ProfileSummaryView from '../../components/summary/ProfileSummaryView';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';
import { summarizeProfile } from '../../profile-logic/summarize-profile';
import { profileSummaryProcessed } from '../../actions/profile-summary';

it('renders ProfileSummaryView correctly', () => {
  /**
   * Mock out any created refs for the components with relevant information.
   */
  function createNodeMock(element) {
    if (element.type === 'div') {
      return {
        offsetWidth: 400,
      };
    }
    return null;
  }

  /* Here we're only interested in the leaf nodes and thus put all
   * samples on the first line.  The samples chosen belong to known
   * categories. */
  const { profile } = getProfileFromTextSamples(`
    pthread_mutex_lock js::RunScript pthread_mutex_lock nsHTMLDNS mozilla::dom::
  `);

  const store = storeWithProfile(profile);

  const profileSummary = renderer.create(
    <Provider store={store}>
      <ProfileSummaryView />
    </Provider>,
    { createNodeMock }
  );
  expect(profileSummary).toMatchSnapshot();

  store.dispatch(profileSummaryProcessed(summarizeProfile(profile)));
  expect(profileSummary).toMatchSnapshot();
});
