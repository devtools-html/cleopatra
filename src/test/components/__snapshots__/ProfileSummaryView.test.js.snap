// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ProfileSummaryView correctly 1`] = `
<div
  className="summarize-profile"
>
  <div
    className="summarize-profile-inner"
  >
    <div>
      <div
        className="summarize-profile-table"
      >
        <div>
          <div
            className="summarize-profile-thread"
            colSpan="3"
          >
            Empty
             Thread, 
            default
             process
          </div>
          <div
            className="summarize-profile-header"
          >
            <div
              className="summarize-line-graph"
              title="The line graph represents the number of samples that were in a given category over time. In order to present a high-level summary, the samples are smoothed with a rolling average."
            >
              Rolling Average of Samples
            </div>
            <div
              className="summarize-profile-details"
            >
              <div
                className="summarize-profile-text"
                title="Each sample's call stack is used to place it in an appropriate category."
              >
                Category
              </div>
              <div
                className="summarize-profile-numeric"
                title="The profiler samples the call stack at a fixed rate. Each of these samples is categorized and then summed over the course of the recording."
              >
                Samples
              </div>
              <div
                className="summarize-profile-numeric"
                title="The percentage of time represents the percentage of how many samples of a given category were observed over the entire course of the recording."
              >
                % Time
              </div>
            </div>
          </div>
        </div>
        <div
          className="summarize-profile-row"
        >
          <div
            className="summarize-line-graph"
          >
            <div
              className="filler summarize-line-graph-filler"
              style={
                Object {
                  "height": "30px",
                }
              }
            />
          </div>
          <div
            className="summarize-profile-details"
          >
            <div
              className="summarize-profile-text"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
            <div
              className="summarize-profile-numeric"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
            <div
              className="summarize-profile-numeric"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
          </div>
        </div>
        <div
          className="summarize-profile-row"
        >
          <div
            className="summarize-line-graph"
          >
            <div
              className="filler summarize-line-graph-filler"
              style={
                Object {
                  "height": "30px",
                }
              }
            />
          </div>
          <div
            className="summarize-profile-details"
          >
            <div
              className="summarize-profile-text"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
            <div
              className="summarize-profile-numeric"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
            <div
              className="summarize-profile-numeric"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
          </div>
        </div>
        <div
          className="summarize-profile-row"
        >
          <div
            className="summarize-line-graph"
          >
            <div
              className="filler summarize-line-graph-filler"
              style={
                Object {
                  "height": "30px",
                }
              }
            />
          </div>
          <div
            className="summarize-profile-details"
          >
            <div
              className="summarize-profile-text"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
            <div
              className="summarize-profile-numeric"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
            <div
              className="summarize-profile-numeric"
            >
              <div
                className="filler summarize-profile-filler"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders ProfileSummaryView correctly 2`] = `
<div
  className="summarize-profile"
>
  <div
    className="summarize-profile-inner"
  >
    <div>
      <div
        className="summarize-profile-table"
      >
        <div>
          <div
            className="summarize-profile-thread"
            colSpan="3"
          >
            Empty
             Thread, 
            default
             process
          </div>
          <div
            className="summarize-profile-header"
          >
            <div
              className="summarize-line-graph"
              title="The line graph represents the number of samples that were in a given category over time. In order to present a high-level summary, the samples are smoothed with a rolling average."
            >
              Rolling Average of Samples
            </div>
            <div
              className="summarize-profile-details"
            >
              <div
                className="summarize-profile-text"
                title="Each sample's call stack is used to place it in an appropriate category."
              >
                Category
              </div>
              <div
                className="summarize-profile-numeric"
                title="The profiler samples the call stack at a fixed rate. Each of these samples is categorized and then summed over the course of the recording."
              >
                Samples
              </div>
              <div
                className="summarize-profile-numeric"
                title="The percentage of time represents the percentage of how many samples of a given category were observed over the entire course of the recording."
              >
                % Time
              </div>
            </div>
          </div>
        </div>
        <div
          className="summarize-profile-row"
        >
          <div
            className="summarize-line-graph"
          >
            <svg
              height={30}
              style={
                Object {
                  "height": "30px",
                  "width": "400px",
                }
              }
              width={400}
            >
              <defs>
                <linearGradient
                  id="summarize-line-graph-gradient"
                  x1="0%"
                  x2="0%"
                  y1="0%"
                  y2="100%"
                >
                  <stop
                    offset="10%"
                    stopColor="#e11800"
                  />
                  <stop
                    offset="60%"
                    stopColor="#56b300"
                  />
                  <stop
                    offset="85%"
                    stopColor="#c8f983"
                  />
                  <stop
                    offset="95%"
                    stopColor="rgba(200, 249, 131, 0)"
                  />
                </linearGradient>
              </defs>
              <path
                d="M-10,0L-10,30M0,1.5L10.256,1.5L82.051,28.5L92.308,28.5L102.564,28.5L112.821,28.5L184.615,1.5L194.872,1.5L205.128,1.5L215.385,1.5L287.179,28.5L297.436,28.5L307.692,28.5L317.949,28.5L389.744,28.5L400,28.5"
                fill="none"
                stroke="url(#summarize-line-graph-gradient)"
                strokeWidth={3}
              />
            </svg>
          </div>
          <div
            className="summarize-profile-details"
          >
            <div
              className="summarize-profile-text"
            >
              wait
            </div>
            <div
              className="summarize-profile-numeric"
            >
              2
            </div>
            <div
              className="summarize-profile-numeric"
            >
              40.0
            </div>
          </div>
        </div>
        <div
          className="summarize-profile-row"
        >
          <div
            className="summarize-line-graph"
          >
            <svg
              height={30}
              style={
                Object {
                  "height": "30px",
                  "width": "400px",
                }
              }
              width={400}
            >
              <defs>
                <linearGradient
                  id="summarize-line-graph-gradient"
                  x1="0%"
                  x2="0%"
                  y1="0%"
                  y2="100%"
                >
                  <stop
                    offset="10%"
                    stopColor="#e11800"
                  />
                  <stop
                    offset="60%"
                    stopColor="#56b300"
                  />
                  <stop
                    offset="85%"
                    stopColor="#c8f983"
                  />
                  <stop
                    offset="95%"
                    stopColor="rgba(200, 249, 131, 0)"
                  />
                </linearGradient>
              </defs>
              <path
                d="M-10,0L-10,30M0,28.5L10.256,28.5L82.051,28.5L92.308,28.5L102.564,28.5L112.821,28.5L184.615,28.5L194.872,28.5L205.128,28.5L215.385,28.5L287.179,28.5L297.436,28.5L307.692,28.5L317.949,28.5L389.744,1.5L400,1.5"
                fill="none"
                stroke="url(#summarize-line-graph-gradient)"
                strokeWidth={3}
              />
            </svg>
          </div>
          <div
            className="summarize-profile-details"
          >
            <div
              className="summarize-profile-text"
            >
              dom
            </div>
            <div
              className="summarize-profile-numeric"
            >
              1
            </div>
            <div
              className="summarize-profile-numeric"
            >
              20.0
            </div>
          </div>
        </div>
        <div
          className="summarize-profile-row"
        >
          <div
            className="summarize-line-graph"
          >
            <svg
              height={30}
              style={
                Object {
                  "height": "30px",
                  "width": "400px",
                }
              }
              width={400}
            >
              <defs>
                <linearGradient
                  id="summarize-line-graph-gradient"
                  x1="0%"
                  x2="0%"
                  y1="0%"
                  y2="100%"
                >
                  <stop
                    offset="10%"
                    stopColor="#e11800"
                  />
                  <stop
                    offset="60%"
                    stopColor="#56b300"
                  />
                  <stop
                    offset="85%"
                    stopColor="#c8f983"
                  />
                  <stop
                    offset="95%"
                    stopColor="rgba(200, 249, 131, 0)"
                  />
                </linearGradient>
              </defs>
              <path
                d="M-10,0L-10,30M0,28.5L10.256,28.5L82.051,28.5L92.308,28.5L102.564,28.5L112.821,28.5L184.615,28.5L194.872,28.5L205.128,28.5L215.385,28.5L287.179,1.5L297.436,1.5L307.692,1.5L317.949,1.5L389.744,28.5L400,28.5"
                fill="none"
                stroke="url(#summarize-line-graph-gradient)"
                strokeWidth={3}
              />
            </svg>
          </div>
          <div
            className="summarize-profile-details"
          >
            <div
              className="summarize-profile-text"
            >
              network
            </div>
            <div
              className="summarize-profile-numeric"
            >
              1
            </div>
            <div
              className="summarize-profile-numeric"
            >
              20.0
            </div>
          </div>
        </div>
        <div
          className="summarize-profile-row"
        >
          <div
            className="summarize-line-graph"
          >
            <svg
              height={30}
              style={
                Object {
                  "height": "30px",
                  "width": "400px",
                }
              }
              width={400}
            >
              <defs>
                <linearGradient
                  id="summarize-line-graph-gradient"
                  x1="0%"
                  x2="0%"
                  y1="0%"
                  y2="100%"
                >
                  <stop
                    offset="10%"
                    stopColor="#e11800"
                  />
                  <stop
                    offset="60%"
                    stopColor="#56b300"
                  />
                  <stop
                    offset="85%"
                    stopColor="#c8f983"
                  />
                  <stop
                    offset="95%"
                    stopColor="rgba(200, 249, 131, 0)"
                  />
                </linearGradient>
              </defs>
              <path
                d="M-10,0L-10,30M0,28.5L10.256,28.5L82.051,1.5L92.308,1.5L102.564,1.5L112.821,1.5L184.615,28.5L194.872,28.5L205.128,28.5L215.385,28.5L287.179,28.5L297.436,28.5L307.692,28.5L317.949,28.5L389.744,28.5L400,28.5"
                fill="none"
                stroke="url(#summarize-line-graph-gradient)"
                strokeWidth={3}
              />
            </svg>
          </div>
          <div
            className="summarize-profile-details"
          >
            <div
              className="summarize-profile-text"
            >
              script
            </div>
            <div
              className="summarize-profile-numeric"
            >
              1
            </div>
            <div
              className="summarize-profile-numeric"
            >
              20.0
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
